spring-cloud-kubernetes-fabric8-config/src/main/java/org/springframework/cloud/kubernetes/fabric8/config/reload/Fabric8EventBasedSecretsChangeDetector.java

```diff
@@ -48,91 +40,74 @@
  */
 public class Fabric8EventBasedSecretsChangeDetector extends ConfigurationChangeDetector {
 
-	private Fabric8SecretsPropertySourceLocator fabric8SecretsPropertySourceLocator;
+	private final Fabric8SecretsPropertySourceLocator fabric8SecretsPropertySourceLocator;
+
+	private final KubernetesClient kubernetesClient;
 
-	private Map<String, Watch> watches;
+	private final boolean monitorSecrets;
 
-	private KubernetesClient kubernetesClient;
+	private SharedIndexInformer<Secret> informer;
 
 	public Fabric8EventBasedSecretsChangeDetector(AbstractEnvironment environment, ConfigReloadProperties properties,
 			KubernetesClient kubernetesClient, ConfigurationUpdateStrategy strategy,
 			Fabric8SecretsPropertySourceLocator fabric8SecretsPropertySourceLocator) {
 		super(environment, properties, strategy);
 		this.kubernetesClient = kubernetesClient;
 		this.fabric8SecretsPropertySourceLocator = fabric8SecretsPropertySourceLocator;
-		this.watches = new HashMap<>();
+		this.monitorSecrets = properties.isMonitoringSecrets();
 	}
 
 	@PreDestroy
-	public void shutdown() {
+	private void shutdown() {
+		if (informer != null) {
+			log.debug("closing secrets informer");
+			informer.close();
+		}
 		// Ensure the kubernetes client is cleaned up from spare threads when shutting
 		// down
-		this.kubernetesClient.close();
+		kubernetesClient.close();
 	}
 
 	@PostConstruct
-	public void watch() {
-		boolean activated = false;
-
-		if (this.properties.isMonitoringSecrets()) {
-			try {
-				activated = false;
-				String name = "secrets-watch-event";
-				this.watches.put(name, this.kubernetesClient.secrets().watch(new Watcher<Secret>() {
-					@Override
-					public void eventReceived(Action action, Secret secret) {
-						if (log.isDebugEnabled()) {
-							log.debug(name + " received event for Secret " + secret.getMetadata().getName());
-						}
-						onEvent(secret);
-					}
-
-					@Override
-					public void onClose(WatcherException exception) {
-						log.warn("Secrects watch closed", exception);
-						Optional.ofNullable(exception).map(e -> {
-							log.debug("Exception received during watch", e);
-							return exception.asClientException();
-						}).map(KubernetesClientException::getStatus).map(Status::getCode)
```

Comment:
I've asked fabric8 team about this `HTTP_GONE` and they told me that with informers this should not be a problem anymore. Still, if we face it, I've let the commented code in the repo. In future, we will know where to look and how to fix.

Reply:
Hmmm was that fixed in a specific k8s version?

Reply:
so [here](https://github.com/fabric8io/kubernetes-client/pull/1800) is where I asked if we should be handling this in case of informers too and where I was told - not really.

Initially, I thought about adding it via the `onNothing` method:

```
@Override
public void onNothing() {
	boolean isStoreEmpty = informer.getStore().list().isEmpty();
	if(!isStoreEmpty) {
		// HTTP_GONE, thus re-inform
		inform();
	}
}
```

Marc said that this should not be needed and I am inclined to trust him, since informers are much more resilient than watchers. I am also inclined to trust this statement, since such a method (where you can track a HTTP_GONE) does not even exist in k8s-native implementation.